tools/pybind11Common.cmake

```diff
@@ -133,6 +134,23 @@ if(MSVC)
   endif()
 endif()
 
+# ----------------------- Optimize binary size --------------------------
+
+add_library(pybind11::opt_size IMPORTED INTERFACE ${optional_global})
+
+if(MSVC)
+  set(PYBIND11_OPT_SIZE /Os)
+else()
+  set(PYBIND11_OPT_SIZE -Os)
+endif()
+
+set_property(
+  TARGET pybind11::opt_size
+  APPEND
+  PROPERTY INTERFACE_COMPILE_OPTIONS $<$<CONFIG:Release>:${PYBIND11_OPT_SIZE}>
+           $<$<CONFIG:MinSizeRel>:${PYBIND11_OPT_SIZE}>
```

Comment:
Is this not already set in the default `MinSizeRel` configuration? I just tried, and I already got this on Linux by just adding `-DCMAKE_BUILD_TYPE=MinSizeRel`.

And how do `-O2` (or `-O3`) and `-Os` then interact? Which concern takes precedence?

I'm wondering if we shouldn't try leaving this to CMake and the user, and add a note to the docs advising about `MinSizeRel`? (Or maybe even making `MinSizeRel` the default build configuration, if nothing is set?)

Reply:
Maybe we could default to MinSizeRel for the tests, and add that as a suggestion to the docs? I do prefer using the CMake way of doing things rather than manually adding flags (plus this way you don't risk having both flags present).

Reply:
It ís optional, this way, so it's not going to be unintentionally harmful to anyone. But CMake might be able to support more exotic compilers/environments?

Reply:
@YannickJadoul -- you are correct to point out that switching to `MinSizeRel` at a global level indeed gives size-based optimizations. The point of this PR is to do so selectively on a per-target level. A very common project layout is:

1. `libmy_project.so` -- all the expensive computation happens in here, optimize as much as possible with auto-vectorization, `-march=native -O3`, etc.
2. `my_project_cli`: linking to `lib_myproject.so` to expose functionality via command line interface
3. `my_project.cpython-..so`: Python bindings. These can become huge (~10-100MB) if the bindings are very comprehensive. Yet, performance here isn't all that critical, because we're just forwarding a bunch of arguments to long-running algorithms that live in the other shared library.

So the goal of this PR is to compile parts (1+2) in release mode, while using MinSizeRel just for the binding target (3).

When specifying a sequence of optimization flags, the latter one have precedence, at least for our compilers in the "officially supported" list.

Reply:
Right, OK, now I see where this is coming from! Thanks for the explanation.
Why not, then :-)